d2graph/copy.go

```diff
@@ -0,0 +1,176 @@
+package d2graph
+
+func (s *Scalar) Copy() *Scalar {
+	if s == nil {
+		return nil
+	}
+	tmp := *s
+	return &tmp
+}
+
+func (s Style) Copy() Style {
+	return Style{
+		Opacity:      s.Opacity.Copy(),
+		Stroke:       s.Stroke.Copy(),
+		Fill:         s.Fill.Copy(),
+		StrokeWidth:  s.StrokeWidth.Copy(),
+		StrokeDash:   s.StrokeDash.Copy(),
+		BorderRadius: s.BorderRadius.Copy(),
+		Shadow:       s.Shadow.Copy(),
+		ThreeDee:     s.ThreeDee.Copy(),
+		Multiple:     s.Multiple.Copy(),
+		Font:         s.Font.Copy(),
+		FontSize:     s.FontSize.Copy(),
+		FontColor:    s.FontColor.Copy(),
+		Animated:     s.Animated.Copy(),
+		Bold:         s.Bold.Copy(),
+		Italic:       s.Italic.Copy(),
+		Underline:    s.Underline.Copy(),
+		Filled:       s.Filled.Copy(),
+	}
+}
+
+func (attrs *Attributes) Copy() *Attributes {
+	if attrs == nil {
+		return nil
+	}
+	return &Attributes{
+		Label:   attrs.Label,
+		Style:   attrs.Style.Copy(),
+		Icon:    attrs.Icon,
+		Tooltip: attrs.Tooltip,
+		Link:    attrs.Link,
+
+		Width:  attrs.Width.Copy(),
+		Height: attrs.Height.Copy(),
+
+		NearKey:  attrs.NearKey,
+		Language: attrs.Language,
+		Shape:    attrs.Shape,
+
+		Direction: attrs.Shape,
+	}
+}
+
+func (o *Object) Copy() *Object {
+	return &Object{
+		Graph:  o.Graph,
+		Parent: o.Parent,
+
+		ID:              o.ID,
+		IDVal:           o.IDVal,
+		Map:             o.Map,
+		LabelDimensions: o.LabelDimensions,
+		References:      append([]Reference(nil), o.References...),
+
+		Box:           o.Box.Copy(),
+		LabelPosition: o.LabelPosition,
+		LabelWidth:    o.LabelHeight,
+		LabelHeight:   o.LabelHeight,
+		IconPosition:  o.IconPosition,
+
+		Class:    o.Class.Copy(),
+		SQLTable: o.SQLTable.Copy(),
+
+		Children:      copyChildrenMap(o.Children),
+		ChildrenArray: append([]*Object(nil), o.ChildrenArray...),
+
+		Attributes: o.Attributes.Copy(),
+
+		ZIndex: o.ZIndex,
+	}
+}
+
+func copyChildrenMap(children map[string]*Object) map[string]*Object {
+	children2 := make(map[string]*Object, len(children))
+	for id, ch := range children {
+		children2[id] = ch
+	}
+	return children2
+}
+
+func (e *Edge) Copy() *Edge {
+	return &Edge{
+		Index: e.Index,
+
+		MinWidth:  e.MinWidth,
+		MinHeight: e.MinHeight,
+
+		SrcTableColumnIndex: e.SrcTableColumnIndex,
+		DstTableColumnIndex: e.DstTableColumnIndex,
+
+		LabelDimensions: e.LabelDimensions,
+		LabelPosition:   e.LabelPosition,
+		LabelPercentage: e.LabelPercentage,
+
+		IsCurve: e.IsCurve,
+		Route:   e.Route,
+
+		Src:          e.Src,
+		SrcArrow:     e.SrcArrow,
+		SrcArrowhead: e.SrcArrowhead.Copy(),
+		Dst:          e.Dst,
+		DstArrow:     e.DstArrow,
+		DstArrowhead: e.DstArrowhead.Copy(),
+
+		References: append([]EdgeReference(nil), e.References...),
+		Attributes: e.Attributes.Copy(),
+
+		ZIndex: e.ZIndex,
+	}
+}
+
+// Copy copies for use as the base of a step or scenario.
+func (g *Graph) Copy() *Graph {
```

Comment:
should we just remove this file since it's unused

Reply:
I'm cool with doing that.

Reply:
(don't forget to remove the reminder comments about adding new fields to Copy)